source/fuzz/transformation_changing_memory_semantics.cpp

```diff
@@ -0,0 +1,296 @@
+                message_.memory_semantics_operand_index() == 1 ? 2 : 3))
+            ->GetSingleWordInOperand(0) &
+        0x1F);
+
+    if (first_5bits_new_memory_semantics == other_memory_semantics_index) {
+      return false;
+    }
+  }
+
+  return true;
+}
+
+void TransformationChangingMemorySemantics::Apply(
+    opt::IRContext* ir_context, TransformationContext* /*unused*/) const {
+  auto needed_atomic_instruction =
+      FindInstruction(message_.atomic_instruction(), ir_context);
+
+  uint32_t needed_index = 0;
+  if (needed_atomic_instruction->opcode() == SpvOpMemoryBarrier) {
+    needed_index = 1;
+  } else {
+    needed_index = message_.memory_semantics_operand_index() == 0 ? 2 : 3;
+  }
+
+  uint32_t higher_bits_memory_semantics =
+      needed_atomic_instruction->GetSingleWordInOperand(needed_index) &
+      0xFFFFFFE0;
+
+  auto new_memory_semantics_instruction = ir_context->get_def_use_mgr()->GetDef(
+      message_.memory_semantics_new_value_id());
+  // New memory semantics value is OR-ed with the second 27bits of old memory
+  // semantics value.
+  uint32_t total_value =
+      new_memory_semantics_instruction->GetSingleWordInOperand(0) |
+      higher_bits_memory_semantics;
+  new_memory_semantics_instruction->SetInOperand(0, {total_value});
+
+  needed_atomic_instruction->SetInOperand(
+      needed_index, {message_.memory_semantics_new_value_id()});
+}
+
+bool TransformationChangingMemorySemantics::IsNeededOpcodeWithAppropriateIndex(
+    SpvOp opcode, uint32_t operand_index) const {
+  switch (opcode) {
+    // Atomic Instructions
+    case SpvOpAtomicLoad:
+    case SpvOpAtomicStore:
+    case SpvOpAtomicExchange:
+    case SpvOpAtomicIIncrement:
+    case SpvOpAtomicIDecrement:
+    case SpvOpAtomicIAdd:
+    case SpvOpAtomicISub:
+    case SpvOpAtomicSMin:
+    case SpvOpAtomicUMin:
+    case SpvOpAtomicSMax:
+    case SpvOpAtomicUMax:
+    case SpvOpAtomicAnd:
+    case SpvOpAtomicOr:
+    case SpvOpAtomicXor:
+    case SpvOpAtomicFlagTestAndSet:
+    case SpvOpAtomicFlagClear:
+    case SpvOpAtomicFAddEXT:
+    // Barrier Instructions
+    case SpvOpControlBarrier:
+    case SpvOpMemoryBarrier:
+    case SpvOpMemoryNamedBarrier:
+      if (operand_index != 0) {
+        return false;
+      }
+      break;
+
+    case SpvOpAtomicCompareExchange:
+    case SpvOpAtomicCompareExchangeWeak:
+      if (operand_index != 0 && operand_index != 1) {
+        return false;
+      }
+      break;
+
+    default:
+      return false;
+  }
+  return true;
+}
+
+bool TransformationChangingMemorySemantics::IsValidConverstion(
+    SpvOp opcode, SpvMemorySemanticsMask first_5bits_old_memory_semantics,
+    SpvMemorySemanticsMask first_5bits_new_memory_semantics,
+    SpvMemoryModel memory_model) {
+  std::set<SpvMemorySemanticsMask> atomic_load_set{
+      SpvMemorySemanticsMaskNone, SpvMemorySemanticsAcquireMask,
+      SpvMemorySemanticsSequentiallyConsistentMask};
```

Comment:
Instead of sets, I suggest adding static functions. E.g. `IsAtomicLoadMemorySemanticsValue(SpvMemorySemanticsMask memory_semantics_value)`. Each function can use a switch case and return true if it is one of the values we expect, otherwise false.

Reply:
Done.